### ISSUE TITLE
Bug of  use BBox3dProjector

### ISSUE THREAD

**Issue by pdoublerainbow** [Owen-Liuyuxuan/visionfactory#1]
Thankes for your open source, I have already see the official Implementation code of monoflex. I have a question about the corner point order, in your code the order is
            [[-1, -1, -1], #0
            [ 1, -1, -1],  #1
            [ 1,  1, -1],  #2
            [ 1,  1,  1],  #3
            [ 1, -1,  1],  #4
            [-1, -1,  1],  #5
            [-1,  1,  1],  #6
            [-1,  1, -1],  #7
            [ 0,  1,  0],  #8
            [ 0, -1,  0],  #9
            [ 0,  0,  0]]  #10,
it looks stranger, in monoflex 
        x_corners = [l / 2, l / 2, -l / 2, -l / 2, l / 2, l / 2, -l / 2, -l / 2]
        y_corners = [0, 0, 0, 0, -h, -h, -h, -h]
        z_corners = [w / 2, -w / 2, -w / 2, w / 2, w / 2, -w / 2, -w / 2, w / 2]
the first four point is top point, next four point is bottom point.  I'm not sure about this, cloud you please check about it.

**Comment by pdoublerainbow** [comment]
You use the same code                 
                keypoints_visible = np.append(
                    np.tile(keypoints_visible[:4] | keypoints_visible[4:8], 2), np.tile(keypoints_visible[8] | keypoints_visible[9], 2)
                ) # "modified keypoint visible from monoflex"
                keypoints_depth_valid = np.stack(
                    (keypoints_visible[[8, 9]].all(), keypoints_visible[[0, 2, 4, 6]].all(), keypoints_visible[[1, 3, 5, 7]].all())
                ).astype(np.float32)
to get keypoints_depth_valid , but the point order in your code is not the same as monoflex.

**Comment by Owen-Liuyuxuan** [comment]
Thanks for your review of the code.

Short Answer:
1. The point order is indeed different from MonoFlex, but it is not a bug.
**2. keypoint_depth_valid  is not used in this repo (and I do not treat it carefully), so it is a bug.** (edited in 2024/2/20, this is a false statement and should be corrected later)

You are correct, the point order is **different** from monoflex. It is the same order as all the codes in visualDet3D/visionfactory (If I remember the history of the codes correctly). 

keypoint_depth_valid is indeed not used in this code repo. Please ignore them first.

The point order in the dataset/BBox3dProject is used to compute depth in "**decode_depth_from_keypoints**". It defined the "group_index" for computing depth from the keypoints.

You could check and make sure that the points 7&0, points 3&4 form group 0 of depth computation; points 2&1, points 6&5 form group 1 of depth computation. This conforms well with the MonoFlex paper (and it is one of the core novel features from the paper that must be implemented correctly to get the training right).

**Comment by pdoublerainbow** [comment]
Thanks for your reply, but keypoints_depth_valid has been used in _keypoints_depth_loss compute
    def _keypoints_depth_loss(self, depths, target, validmask, uncertainty):

        loss = F.l1_loss(depths, target.repeat(1, 3), reduction='none') * torch.exp(-uncertainty) + \
            uncertainty * self.uncertainty_weight #[N, 3]
        
        valid_loss = loss * validmask.float() + (1 - validmask.float()) * loss.detach() #[N, 3]

        return valid_loss.mean(dim=1).sum() / (len(depths) + 1e-4)
In my view, 
                keypoints_visible = np.append(
                    np.tile(keypoints_visible[:4] | keypoints_visible[4:8], 2), np.tile(keypoints_visible[8] | keypoints_visible[9], 2)
                ) # "modified keypoint visible from monoflex"
                keypoints_depth_valid = np.stack(
                    (keypoints_visible[[8, 9]].all(), keypoints_visible[[0, 2, 4, 6]].all(), keypoints_visible[[1, 3, 5, 7]].all())
                ).astype(np.float32)
This part of code should be corrected, am i right?

**Comment by pdoublerainbow** [comment]
I change the code look like this 
`                keypoints_visible = np.append(
                    keypoints_visible[:8] | keypoints_visible[[7,2,1,4,3,6,5,0]], np.tile(keypoints_visible[8] | keypoints_visible[9], 2)
                ) # "modified keypoint visible from monoflex"
                keypoints_depth_valid = np.stack(
                    (keypoints_visible[[8, 9]].all(), keypoints_visible[[0, 3, 4, 7]].all(), keypoints_visible[[1, 2, 5, 6]].all())
                ).astype(np.float32)
`

**Comment by Owen-Liuyuxuan** [comment]
Thank you so much for noticing this bug. I did not understand this part carefully at the time.

I read your code here and I believe you made it correct. (declare depth visible if either the top or the bottom point is indeed visible).

Would you mind forking the repo and making a PR to the repo to show that you are the main author of this update?

**Comment by pdoublerainbow** [comment]
Hi, I have maked aPR to this repo.
